### PR TITLE
Generative VQVAE Model for the OASIS Dataset - s4698512

### PR DESCRIPTION
**Pull request by hugs7**
**Name:** Hugo Burton
**Student Number:** s4698512

This pull request implements a generative model of the OASIS dataset using a Vector Quantised Variational Auto Encoder. Please see the README file for more details regarding the model, implementation, usage and performance.

**Task:** Create a generative model of one of the OASIS brain, ADNI brain or OAI AKOA knee data set using a VQVAE or VQVAE2 that has a "reasonably clear image" and a Structured Similarity (SSIM) of over 0.6. [Hard Difficulty]

**Model:** A VQVAE model was implemented in Pytorch with 3 convolutional layers plus an additional two ResBlock layers in both the encoder and decoder, each with a hidden layer count of 128. Input and output images and 256x256 greyscale. A codebook with 32 latent embeddings was used in between the encoder and decoder.

**Usage:** Download the OASIS dataset and place it in a folder ../PatternAnalysisData/datasets/OASIS/. Ensure you are running in a conda environment with PyTorch installed. Some other dependencies are also required as documented in the README.

This pull request also features support for the ADNI dataset. Simply change the dataset variable in the train and predict modules to switch.

**Performance:** The model achieved a structured similarity of 0.849 on the test data.

**Files Added:**

1. `README.md` Report and overview of the VQVAE model, implementation, usage, results and performance
2. `dataset.py` Loads the OASIS and ADNI datasets and performs pre-processing
3. `modules.py` Contains all the model structure for the VQVAE as well as helper functions to handle backpropagation
4. `train.py` Driver script for training on the model
5. `predict.py` Reads a save file for the model and generates sample images as well as plots performance metrics about the model
6. `./Assets` figures and other data used in the README file.

### CODE REVIEW

**Comment by SiyuLiu0329** [comment]
**This is an initial inspection, no action is required at this point**

Difficulty: Hard

Readme:
- Project Overview: Good
- Model: Good
- Data + Preprocessing: Good
- Usage: Good
- Training / Loss Curve(s): Good
- Result Demonstration: Good, but no image generation results
- References: Good

Functionality:
- Reconstruction: Good
- Generation: **Not completed**

Code:
- Consistent with Results: Yes
- File Structure: Good
- Commenting: Good but could be a bit more descriptive
- Commit frequency: Good
- Commit messages: Good

Other Comments:
N/A

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks
## Recognition Problem 
Solves problem (no generation) -1
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage (no generation results) -1
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe acceptable
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
